### PR TITLE
Honor --force option in configurators

### LINKED ISSUE #457: sync:recipes --force does not update .gitignore entries
When testing the `sync:recipes --force`, my `.gitignore` file WAS updated to add *new* packages that previously did not have any `.gitignore` rules and now have some.

But, for FrameworkBundle (for example), I DID already have some entries in `.gitignore` from its recipe. However, those entries were *not* updated.

To replicate

```
git clone git@github.com:symfony/symfony-demo.git
cd symfony-demo
git checkout v1.2.7
composer install
rm -rf vendor/symfony/flex
git clone git@github.com:symfony/flex.git vendor/symfony/flex
composer sync-recipes --force
```

### PR DESCRIPTION
**Pull request by renanbr**
Fixes #457

This PR is a continuity of https://github.com/symfony/flex/pull/458

### CODE REVIEW

**Comment by renanbr** [comment]
Updates:

- `GitignoreConfigurator`, `MakefileConfigurator`, `EnvConfigurator` are impacted
- They preserve the order of sections
- `AbstractConfigurator` has a new method called `reconfigure(string $file, string $data)`
    - markers are guessed based on the first and last non-empty lines of `$data` given
    - it is the configurator's responsability to build the contents

File: src/Configurator/AbstractConfigurator.php [hunk 1]
```diff
@@ -69,4 +69,24 @@ protected function markXmlData(Recipe $recipe, string $data): string
...
+    /**
+     * @return bool True if section was found and replaced
+     */
+    protected function updateData(string $file, string $data): bool
+    {
+        $pieces = explode("\n", trim($data));
+        $startMark = trim(reset($pieces));
+        $endMark = trim(end($pieces));
+        $contents = file_get_contents($file);
+        if (false === strpos($contents, $startMark) || false === strpos($contents, $endMark)) {
```
Comment (nicolas-grekas) [inline src/Configurator/AbstractConfigurator.php]: may I suggest a blank line just before the "if"

File: src/Configurator/AbstractConfigurator.php [hunk 2]
```diff
@@ -69,4 +69,24 @@ protected function markXmlData(Recipe $recipe, string $data): string
...
+            return false;
+        }
+
+        $pattern = '/'.preg_quote($startMark, '/').'.*'.preg_quote($endMark, '/').'/s';
```
Comment (nicolas-grekas) [inline src/Configurator/AbstractConfigurator.php]: `s/.*/.*?/` (ungreeded regexp would be better)

File: src/Configurator/EnvConfigurator.php [hunk 3]
```diff
@@ -22,9 +22,9 @@ public function configure(Recipe $recipe, $vars, array $options = [])
     {
         $this->write('Added environment variable defaults');
 
-        $this->configureEnvDist($recipe, $vars);
+        $this->configureEnvDist($recipe, $vars, $options);
```
Comment (nicolas-grekas) [inline src/Configurator/EnvConfigurator.php]: `$this->configureEnvDist($recipe, $vars, $options['force'] ?? false);`

File: src/Configurator/EnvConfigurator.php [hunk 4]
```diff
@@ -22,9 +22,9 @@ public function configure(Recipe $recipe, $vars, array $options = [])
...
         if (!file_exists($this->options->get('root-dir').'/.env.test')) {
-            $this->configurePhpUnit($recipe, $vars);
+            $this->configurePhpUnit($recipe, $vars, $options);
```
Comment (nicolas-grekas) [inline src/Configurator/EnvConfigurator.php]: same

File: src/Configurator/EnvConfigurator.php [hunk 5]
```diff
@@ -34,15 +34,15 @@ public function unconfigure(Recipe $recipe, $vars)
         $this->unconfigurePhpUnit($recipe, $vars);
     }
 
-    private function configureEnvDist(Recipe $recipe, $vars)
+    private function configureEnvDist(Recipe $recipe, $vars, array $options)
```
Comment (nicolas-grekas) [inline src/Configurator/EnvConfigurator.php]: `configureEnvDist(Recipe $recipe, $vars, bool $update)` (or `bool $force`)

File: src/Configurator/EnvConfigurator.php [hunk 6]
```diff
@@ -34,15 +34,15 @@ public function unconfigure(Recipe $recipe, $vars)
...
     {
         foreach (['.env.dist', '.env'] as $file) {
             $env = $this->options->get('root-dir').'/'.$file;
             if (!is_file($env)) {
                 continue;
             }
 
-            if ($this->isFileMarked($recipe, $env)) {
+            if ($this->isFileMarked($recipe, $env) && empty($options['force'])) {
```
Comment (nicolas-grekas) [inline src/Configurator/EnvConfigurator.php]: `if (!$update && $this->isFileMarked($recipe, $env)) {`

File: src/Configurator/EnvConfigurator.php [hunk 7]
```diff
@@ -62,19 +62,22 @@ private function configureEnvDist(Recipe $recipe, $vars)
...
-            file_put_contents($env, $data, FILE_APPEND);
+
+            if (!$this->updateData($env, $data)) {
+                file_put_contents($env, $data, FILE_APPEND);
+            }
         }
     }
 
-    private function configurePhpUnit(Recipe $recipe, $vars)
+    private function configurePhpUnit(Recipe $recipe, $vars, array $options)
```
Comment (nicolas-grekas) [inline src/Configurator/EnvConfigurator.php]: same (etc.)

File: src/Configurator/EnvConfigurator.php [hunk 8]
```diff
@@ -62,19 +62,22 @@ private function configureEnvDist(Recipe $recipe, $vars)
...
     {
         foreach (['phpunit.xml.dist', 'phpunit.xml'] as $file) {
             $phpunit = $this->options->get('root-dir').'/'.$file;
             if (!is_file($phpunit)) {
                 continue;
             }
 
-            if ($this->isFileXmlMarked($recipe, $phpunit)) {
+            if ($this->isFileXmlMarked($recipe, $phpunit) && empty($options['force'])) {
```
Comment (nicolas-grekas) [inline src/Configurator/EnvConfigurator.php]: `if (!$update && $this->isFileXmlMarked($recipe, $phpunit)) {`

File: src/Configurator/GitignoreConfigurator.php [hunk 9]
```diff
@@ -23,7 +23,7 @@ public function configure(Recipe $recipe, $vars, array $options = [])
         $this->write('Added entries to .gitignore');
 
         $gitignore = $this->options->get('root-dir').'/.gitignore';
-        if ($this->isFileMarked($recipe, $gitignore)) {
+        if ($this->isFileMarked($recipe, $gitignore) && empty($options['force'])) {
```
Comment (nicolas-grekas) [inline src/Configurator/GitignoreConfigurator.php]: same (etc)

**Comment by renanbr** [comment]
tweaks applied, squashed

it's ready ready to be reviewed

**Comment by nicolas-grekas** [comment]
Thank you @Pierstoval.

**Comment by nicolas-grekas** [comment]
Thank you 469

**Comment by nicolas-grekas** [comment]
Thank you @renanbr!

**Pull request merged**